### PR TITLE
Fleet rvalue

### PR DESCRIPTION
**Pull request by UnorderedSigh**
conditions in fleets

### CODE REVIEW

**Comment by warp-core** [comment]
Is three words really an adequate descriptions?

**Comment by Amazinite** [comment]
Use the template.

File: source/Fleet.h [hunk 1]
```diff
@@ -82,13 +89,14 @@ class Fleet {
...
+	bool hasConditions = true;
 	std::string fleetName;
 	const Government *government = nullptr;
 	const Phrase *names = nullptr;
 	const Phrase *fighterNames = nullptr;
-	WeightedList<Variant> variants;
+	WeightedList<Variant,WeightType> variants;
 	// The number of different items the ships in this fleet will carry in cargo.
-	int cargo = 3;
+	RValue<int> cargo;
```
Comment (rovermicrover) [inline source/Fleet.h:99]: What advantages do you see wrapping the cargo int in this RValue class give you here?

File: source/RValue.h [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+#include <stdexcept>
+#include <string>
+
+
+#define DEBUG_RVALUE_CONDITIONS
+
+
+
+// This class stores an rvalue and remembers where the rvalue came
```
Comment (rovermicrover) [inline source/RValue.h:26]: It seems odd to give and int or a string special meaning depending on its source. Seems like that is just a way to smuggle context, and then by doing so couple that context to what ever method uses the r value key to do conditional logic.

File: source/RandomEvent.h [hunk 3]
```diff
@@ -17,42 +17,74 @@ this program. If not, see <https://www.gnu.org/licenses/>.
...
+	// Update the period from a ConditionSet if needed:
+	template<typename Getter>
+	void UpdateConditions(const Getter &getter);
+
+	bool HasConditions() const { return PeriodTypeHasConditions<P>::HasConditions(period); }
 
 private:
 	const T *event;
-	int period;
+	P period;
```
Comment (rovermicrover) [inline source/RandomEvent.h:58]: Why this change, it seems like you just want to use your RValue here wrapping and int.

File: source/System.cpp [hunk 4]
```diff
@@ -426,6 +426,15 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
 	// Systems without an asteroid belt defined default to a radius of 1500.
 	if(belts.empty())
 		belts.emplace_back(1, 1500.);
+
+	// Update cached info about whether there are conditions
```
Comment (rovermicrover) [inline source/System.cpp:430]: Why the need for a cache here? What computational advantage does caching here give you?

**Review by rovermicrover — COMMENTED** [review]
Did you mean to set this up as a draft or against your own branch?

Have some questions just for my information, I have no approval power.

**Comment by UnorderedSigh** [comment]
Sorry! This went to the wrong fork!

**Comment by UnorderedSigh** [comment]
This is quite embarrassing. The code was half-written, and RValue still sounds silly to me. (Literally everything in your code can be an RValue, unless you're assigning to it.)

File: source/System.cpp (hunk 4 above)
Comment (UnorderedSigh) [reply source/System.cpp:430]: All Systems and Fleets have to be processed every time an event is applied, the player lands, the player departs, or a save is loaded. That adds up quickly unless the game knows not to recurse into the unaffected Systems and Fleets. Very few (if any) will have conditions in them.

File: source/Fleet.h (hunk 1 above)
Comment (UnorderedSigh) [reply source/Fleet.h:99]: For anything to be configurable via a condition, it needs an RValue so you can map it back to the condition variable name it came from. Luckily, cargo is an int, not a double, so it can be configured via condition.

File: source/RValue.h (hunk 2 above)
Comment (UnorderedSigh) [reply source/RValue.h:26]: Yes. You need the context so you can reapply the condition variable after it updates.

File: source/RandomEvent.h (hunk 3 above)
Comment (UnorderedSigh) [reply source/RandomEvent.h:58]: Yes. That should not have been an int. Thank you for catching that.

**Comment by warp-core** [comment]
> Sorry! This went to the wrong fork!

Been there.
